File: app/schemas/questions.py
from datetime import datetime

from pydantic import BaseModel, Field


class Answer(BaseModel):
    text: str = Field(..., max_length=1024)


class FullAnswer(Answer):
    is_valid: bool = Field(default=False, alias="isValid")


class AnswerResul(Answer):
    is_valid: bool = Field(default=False, alias="isValid")
    true_valid: bool = Field(default=False, alias="trueValid")


class Question(BaseModel):
    text: str = Field(..., max_length=2048)
    answers: list[Answer]
    image: str = Field(default="", max_length=1024)


class ValidateQuestion(Question):
    answers: list[FullAnswer]


class FullQuestion(Question):
    answers: list[FullAnswer]
    explanation: str = Field(default="", max_length=2048)


class QuestionResult(FullQuestion):
    answers: list[AnswerResul]


class QuestionGroupToPass(BaseModel):
    id: str = Field(..., alias="_id")
    name: str
    questions: list[Question]
    description: str
    completion_time_seconds: int = Field(
        default=0,
        description="Время для прохождения теста",
        alias="completionTimeSeconds",
    )


class UpdateQuestionGroup(BaseModel):
    name: str = Field(..., max_length=256)
    tags: list[str]
    questions: list[FullQuestion]
    description: str = Field(default="", max_length=2048)
    timeout_minutes: int = Field(
        default=0,
        description="Время для повторного прохождения",
        alias="timeoutMinutes",
    )
    completion_time_minutes: int = Field(
        default=0,
        description="Время для прохождения теста",
        alias="completionTimeMinutes",
    )


class CreateQuestionGroup(UpdateQuestionGroup):
    pass


class FullQuestionGroup(BaseModel):
    id: str = Field(..., alias="_id")
    name: str
    user_id: str = Field(..., alias="userId")
    tags: list[str]
    questions: list[FullQuestion]
    description: str
    timeout_minutes: int = Field(
        default=0,
        description="Время для повторного прохождения",
        alias="timeoutMinutes",
    )
    completion_time_minutes: int = Field(
        default=0,
        description="Время для прохождения теста",
        alias="completionTimeMinutes",
    )
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")


class ValidateQuestionGroup(BaseModel):
    id: str = Field(..., alias="_id")
    questions: list[ValidateQuestion]


class MinimalQuestionGroup(BaseModel):
    id: str = Field(..., alias="_id")
    name: str
    user_id: str = Field(..., alias="userId")
    username: str
    description: str
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")
    timeout_minutes: int = Field(
        ...,
        description="Время для повторного прохождения",
        alias="timeoutMinutes",
    )
    completion_time_minutes: int = Field(
        ...,
        description="Время для прохождения теста",
        alias="completionTimeMinutes",
    )
    tags: list[str]


class QuestionGroupResult(BaseModel):
    id: str = Field(..., alias="_id")
    name: str
    tags: list[str]
    description: str
    timeout_minutes: int = Field(
        ..., description="Время для повторного прохождения", alias="timeoutMinutes"
    )
    completion_time_minutes: int = Field(
        ..., description="Время для прохождения теста", alias="completionTimeMinutes"
    )
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")
    questions: list[QuestionResult]
    total_score: int = Field(..., alias="totalScore")
    user_score: int = Field(..., alias="userScore")

File: frontend/src/questions.ts
class Answer {
    constructor(
        public text: string = "",
        public isValid: boolean = false,
        public trueValid: boolean = true,
    ) {}

    public get isCorrect() {
        if (this.trueValid === undefined) return true
        return !this.isValid || this.trueValid === this.isValid
    }
}

class Question {
    constructor(
        public text: string = "",
        public answers: Array<Answer> = [new Answer()],
        public image: string = "",
        public explanation: string = "",
    ) {}

    addAnswer(): void {
        this.answers.push(new Answer())
    }

    deleteAnswer(answerID: number): void {
        this.answers.splice(answerID, 1)
    }
}


class TestAbout {
    constructor(
        public _id: string,
        public name: string,
        public description: string,
        public username: string,
        public completionTimeMinutes: number,
        public timeoutMinutes: number,
        public tags: Array<string>
    ) {}
}


class FullTest extends TestAbout {
    questions: Array<Question>
    constructor(name: string,
                description: string,
                completionTimeMinutes: number,
                timeoutMinutes: number,
                tags: Array<string>,
                questions: Array<Question>
    ) {
        super("", name, description, null, completionTimeMinutes, timeoutMinutes, tags);
        this.questions = questions
    }

    addQuestion(): void {
        this.questions.push(
            new Question(
                "",
                [
                    new Answer("Ответ 1", false),
                    new Answer("Верный ответ", true)
                ],
                "",
                ""
            )
        )
    }

    deleteQuestion(questionID: number): void {
        this.questions.splice(questionID, 1)
    }
}


class TestResultMixin {

    constructor(
        public userScore: number = null,
        public totalScore: number = null
    ) {}

    public get percents(): number {
        return Math.floor((this.userScore / this.totalScore) * 100)
    }

    public get passedSuccessfully(): boolean {
        return this.userScore / this.totalScore > 0.7
    }
}


class TestForPassing extends TestResultMixin {
    constructor(
        public _id: string,
        public name: string,
        public description: string,
        public completionTimeSeconds: number,
        public questions: Array<Question>,
        public userScore: number,
        public totalScore: number,
    ) {
        super(userScore, totalScore)
    }

}

class PassedQuestion extends TestResultMixin {
    createdAt: Date
    questionGroupId: string
    questionGroupName: string
    totalScore: number
    userId: string
    userScore: number
    constructor(createdAt: string, questionGroupId: string, questionGroupName: string,
                totalScore: number, userId: string, userScore: number) {
        super(userScore, totalScore)
        this.createdAt = new Date(createdAt)
        this.questionGroupId= questionGroupId
        this.questionGroupName= questionGroupName
        this.userId= userId
    }

    public get dateString(): string {
        return new Intl.DateTimeFormat(
            "ru",
            {day: "numeric", month: "long", year: "numeric", hour: "numeric", minute: "numeric"}
        ).format(this.createdAt)
    }

}

function createNewQuestions(data: Array<any>): Array<Question> {
    let questions: Array<Question> = []
    for (const questionData of data) {
        let answers: Array<Answer> = []
        for (const answerData of questionData.answers) {
            answers.push(new Answer(answerData.text, answerData.isValid, answerData.trueValid))
        }
        questions.push(new Question(questionData.text, answers, questionData.image, questionData.explanation))
    }
    return questions
}


function createNewTestForPassing(data: any): TestForPassing {
    return new TestForPassing(
        data._id,
        data.name,
        data.description,
        data.completionTimeSeconds,
        createNewQuestions(data.questions),
        data.userScore,
        data.totalScore,
    )
}

function createNewFullTest(data: any): FullTest {
    return new FullTest(
        data.name,
        data.description,
        data.completionTimeMinutes,
        data.timeoutMinutes,
        data.tags,
        createNewQuestions(data.questions),
    )
}


function createNewTestAboutList(data: Array<any>): Array<TestAbout> {
    let res: Array<TestAbout> = []
    for (const testData of data) {
        res.push(
            new TestAbout(testData._id, testData.name, testData.description, testData.username,
                testData.completionTimeMinutes, testData.timeoutMinutes, testData.tags)
        )
    }
    return res
}

export {
    TestAbout,FullTest,Question,Answer,PassedQuestion,TestForPassing,
    createNewTestAboutList,createNewFullTest,createNewTestForPassing
}
